sound/soc/sof/amd/acp.c

```diff
@@ -535,6 +630,28 @@ int amd_sof_acp_probe(struct snd_sof_dev *sdev)
 	if (ret < 0)
 		goto free_ipc_irq;
 
+	/* scan SoundWire capabilities exposed by DSDT */
+	ret = acp_sof_scan_sdw_devices(sdev, chip->sdw_acpi_dev_addr);
+	if (ret < 0) {
+		dev_dbg(sdev->dev, "skipping SoundWire, not detected with ACPI scan\n");
+		goto skip_soundwire;
+	}
+	ret = amd_sof_sdw_probe(sdev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: SoundWire probe error\n");
+		free_irq(sdev->ipc_irq, sdev);
+		pci_dev_put(adata->smn_dev);
+		return ret;
+	}
+
+	ret = amd_sof_sdw_start(sdev);
```

Comment:
so this answers to my question: since you can do the probe and start in a single step, you can collapse all the work fron in start() in probe().

This was different for Intel because we had to defer the start() to post firmware-run

Reply:
We can't combine start() in probe sequence reason as follows.
Probe () function returns sdw_amd_ctx  for the platform devices created which is serialized. if we pull the start() code probe() , Once the link gets started, Soundwire interrupt reported in ACP PCI driver and it gives kernel panic as we are referring pdev from sdw context in parent driver interrupt handler.

We want to serialize active manager links probe() sequence followed by start() sequence. This sequence is getting called before loading the firmware in our case.

Reply:
sorry, what?

Your code does 
````c
ret = amd_sof_sdw_probe(sdev);
...
ret = amd_sof_sdw_start(sdev);
````
so they are called serially and there's nothing in between. I don't understand why you cannot collapse them into a single step. If you have some sort of timing issue, then the code above is racy as hell and certainly wrong.

Please re-visit this...

Reply:
we misinterpreted your earlier comment. We will make the changes.